### ISSUE TITLE
empty attributes after save()

### ISSUE THREAD

**Issue by couchbelag** [flexirest/flexirest#151]
### Expected behaviour

I followed the tutorials to implement flexirest in our rails application, trying to get the same behavior as with ActiveRecord (create, find, save,...). I have to say it's absolutely amazing how fast we implemented this with this gem! Great job!

It works like a charm, but there is one thing I don't understand... or maybe I'm doing it all wrong. In our scenario the server will respond to e.g. `save()` with 200/ok and **empty** body. Hence after `save()` the attributes hash is empty, no way to restore it. We even lost the ID to `find()` the object from the server again.

### Desired behaviour

I'd expect the attributes to be remembered even if a server won't respond with attribute in body again. So I'd assume attributes from original object would be merged with attributes received in response. I worked around this by overwriting `Flexirest::BaseWithoutValidation#_copy_from()` in my own class:

```ruby
class MyResource < Flexirest::Base
  base_url "..."

  patch :save, "/:id", only_changed: true

  def _copy_from(result)
    backup = @attributes
    super(result)
    @attributes = backup.merge(@attributes)
  end
end
```

But I'm not sure if this is the proper way to do this. Or am I missing some hooks or switches that would result in this behavior?

### How to reproduce

Hard to reproduce, as you need to find/implement a server not responding with the object if it is updated. If you have one, it's simply read-modify-write:

```ruby
r = MyResource.find("res-4711")
r.some_attribute = "new value"
r.save
# => r._attributes is empty
```

**Comment by andyjeffries** [comment]
First off, thanks for the compliments, I'm glad you found this gem useful!

Secondly, thanks for reporting this, it's an interesting issue. 

I think it's actually a combination of an issue in Flexirest AND an issue in your API. Most RESTful guides I've seen say that a 200 OK should contain a representation of the object, and if that's not the case, it should return a 204 No Content. For example: https://restfulapi.net/http-status-200-ok/ (although it implies an empty payload is OK, but then says about the 204).

However, I think even if Flexirest gets a 204, it will still lose the current attributes. So I need to fix that first of all so the gem implements best practices/correct behaviour.

Then we come to your specific API. I'll assume for now that you can't (or won't) change that to return a 204. So then I think we need an option (maybe per class, or maybe per method) to say "ignore_empty_responses" to handle this case and not wipe the internal attributes if a non-204 is returned and the body is empty.

I'll get on to those.

**Comment by andyjeffries** [comment]
Fixed and documented how to use it at https://github.com/flexirest/flexirest/blob/master/docs/empty-body-handling.md

Available now as v1.10.9.

Let me know if you have any further issues.

**Comment by couchbelag** [comment]
I'm impressed, 16 hours from issue report to new release! Not to mention that we're most propably living in differen time zones. Guys, I love this project! :+1: 

`ignore_empty_response` is working perfectly for me, thanks! And you're assumption is right, I suppose I won't be able to convince people managing the servers to change that REST API. 

I still had another thought related to this issue... I remember REST APIs that actually do respond with contents in body, but it could be status information only, e.g. `{"result": true}`. So allthough it's a 200-response attributes might be wiped. Well, I don't think this should be covered in this issue's context (1.10.9 is working for my use case anyway), but maybe it's some feature for future releases? I could open up another issue describing this in detail again, if you wish? On the other hand let's not encourage people to implement such APIs :grin:

**Comment by andyjeffries** [comment]
Thank you very much :-)

I'm lucky in my current role the web and API teams report to me, so I can force through any changes, but I've been around long enough to know not everyone is that lucky ;-)

I think the current proxying feature https://github.com/flexirest/flexirest/blob/master/docs/proxying-apis.md would handle that case. If it wiped out the body completely then the new `ignore_empty_response` handler could take over. However, that would rely on a `"result": false` actually coming with a different HTTP status code, not just still with a 200.

Let's leave it for now though, if someone asks for this because they have that situation then I'll have a re-think.

If you have any other feature requests, definitely open a new issue though, I appreciate them all.

**Comment by andyjeffries** [comment]
By the way, this morning I released 1.10.10. It doesn't really affect anything, but just ensures that a 1.x branch of Faraday is required, the 2.x releases removed engines and re-including them as they suggest doesn't work for me. Anyway, may want to upgrade it to stop others on your project doing a blind `bundle update` and finding stuff breaks.

**Comment by couchbelag** [comment]
> Let's leave it for now though, if someone asks for this because they have that situation then I'll have a re-think.

I fully agree, stick to "agile" :grin: 
And thanks again for the hint regarding 1.10.10, just upgraded! Guess it would have taken ages to figure out...
